File: GPLDicToModel-All/OCDicToModel/OCDicToModel/HttpRequest/HttpServiceEngine.m
//
//  HttpServiceEngine.m
//  OCDicToModel
//
//  Created by GPL on 2017/12/21.
//  Copyright © 2017年 GPL. All rights reserved.
//

#import "HttpServiceEngine.h"

typedef NS_ENUM(NSInteger, RequestKindStyle) {
    RequestKindStylePost = 0,
    RequestKindStylePostJson,
    RequestKindStyleGet,
};

static NSInteger const HttpRequestTimeOut = 30;

static HttpServiceEngine *serviceEngine = nil;

@interface HttpServiceEngine()
@property (nonatomic, strong) NSURLSession *session;
@end

@implementation HttpServiceEngine
+ (instancetype)sharedInstance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        serviceEngine = [[HttpServiceEngine alloc] init];
    });
    return serviceEngine;
}
- (instancetype)init {
    self = [super init];
    if (self) {
        self.session = [NSURLSession sharedSession];
    }
    return self;
}

/**
 GET请求方式
 
 @param request 请求体
 */
- (void)asyncGetRequest:(BaseRequest *)request {
    [self asyncRequest:request requestStyle:RequestKindStyleGet];
}

/**
 POST JSON 请求方式
 
 @param request 请求体
 */
- (void)asyncPostJsonRequest:(BaseRequest *)request {
    [self asyncRequest:request requestStyle:RequestKindStylePostJson];
}
//通用请求体
- (void)asyncRequest:(BaseRequest *)request requestStyle:(RequestKindStyle)style {
    if (request) {
        dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
        dispatch_async(queue, ^{
            NSString *url = [request url];
            NSLog(@"reqest url = %@", url);
            if (url == nil || url.length < 1) {
                NSLog(@"[%@]请求URL出错了", [request class]);
                return ;
            }
            NSMutableURLRequest *urlRequest;
            if (style == RequestKindStylePost) {

                NSData *data = [request toPostJsonData];
                NSURL *URL = [NSURL URLWithString:url];
                urlRequest = [NSMutableURLRequest requestWithURL:URL];

                [urlRequest setTimeoutInterval:HttpRequestTimeOut];
                if (data) {
                    [urlRequest setValue:[NSString stringWithFormat:@"%ld",(long)[data length]] forHTTPHeaderField:@"Content-Length"];
                    [urlRequest setHTTPMethod:@"POST"];
                    [urlRequest setValue:@"application/json; charset=UTF-8" forHTTPHeaderField:@"Content-Type"];
                    [urlRequest setValue:@"gzip" forHTTPHeaderField:@"Accept-Encoding"];
                    [urlRequest setHTTPBody:data];
                }
                
            }else if (style == RequestKindStyleGet){
                
                NSURL *URL = [NSURL URLWithString:[request toGetUrlString]];
                urlRequest = [NSMutableURLRequest requestWithURL:URL];
                
                [urlRequest setTimeoutInterval:HttpRequestTimeOut];
                
                [urlRequest setCachePolicy:NSURLRequestReloadIgnoringCacheData];
                [urlRequest setHTTPShouldHandleCookies:FALSE];
                [urlRequest setHTTPMethod:@"GET"];

            }
            
            NSURLSessionDataTask *task = [_session dataTaskWithRequest:urlRequest completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
                if (error != nil) {
                    if (request.failBlock) {
                        dispatch_async(dispatch_get_main_queue(), ^{
                            request.failBlock(request);
                        });
                    }else {
                        NSLog(@"请求失败[%@]",error.localizedFailureReason);
                    }
                }else {
                    NSString *responseString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
                    NSLog(@"[%@] 请求的返回原数据串 :\n================================\n %@ \n================================" , [request class], responseString);
                    NSObject *jsonObj = [NSJSONSerialization JSONObjectWithData:data options:0 error:NULL];
                    [request parseResponse:jsonObj];
                }
            }];
            [task resume];
        });
    }
}
@end

File: GPLDicToModel-All/SwiftJsonToModel/SwiftJsonToModel/HttpRequest/HttpServiceEngine.swift
//
//  HttpServiceEngine.swift
//  SwiftJsonToModel
//
//  Created by 顾鹏凌 on 2017/12/22.
//  Copyright © 2017年 顾鹏凌. All rights reserved.
//

import UIKit
enum HttpKindStyle : NSString {
    case Post = "Post"
    case PostJson = "PostJson"
    case Get = "Get"
}
let HttpRequestTimeOut:Int = 30

class HttpServiceEngine: NSObject {
    static let shareInstance = HttpServiceEngine()
    var session:URLSession = URLSession.shared
    
    func asyncGetRequest(req:BaseRequest) {
        self.asyncRequest(req: req, style: .Get)
    }
    func asyncPostJsonRequest(req:BaseRequest) {
        self.asyncRequest(req: req, style: .PostJson)
    }
    func asyncRequest(req:BaseRequest,style:HttpKindStyle) {
        DispatchQueue.global(qos: .default).async {
            let url = req.url()
            print(">>>>>请求接口\(url)")
            if url.lengthOfBytes(using: String.Encoding.utf8) < 1 {
                print(">>>>>请求url错误\(url)")
                return
            }
            
            let request:NSMutableURLRequest = NSMutableURLRequest.init()
            if style == .PostJson {
                request.url = NSURL.init(string: url)! as URL
//                request = NSMutableURLRequest.init(url: NSURL.init(string: url)! as URL)
                let data:NSData = (req.toPostJsonData())! as NSData
                
                request.httpMethod = "POST"
                request.setValue("application/json; charset=UTF-8", forHTTPHeaderField: "Content-Type")
                request.setValue("gzip", forHTTPHeaderField: "Accept-Encoding")
                request.httpBody = data as Data

                
            }else if style == .Get {
//                request = NSMutableURLRequest.init(url: NSURL.init(string: req.toGetUrlString())! as URL)
                request.url = NSURL.init(string: req.toGetUrlString())! as URL

                request.httpShouldHandleCookies = false
                request.httpMethod = "GET"
            }
            request.timeoutInterval = TimeInterval(HttpRequestTimeOut)
            
            print(">>>>>正在请求接口\(style)")

            let dataTask:URLSessionDataTask = self.session.dataTask(with: request as URLRequest, completionHandler: { (data, response, error) in
                print(">>>>>请求接口完成\n")
                if (data != nil) {
                    let responseData = try! JSONSerialization.jsonObject(with: data!, options: .allowFragments)
                    req.parseResponse(respJsonObject: responseData as AnyObject)
                }else {
                    DispatchQueue.main.async(execute: {
                        req.resopnse.error = -1
                        req.resopnse.message = "返回数据非Json格式"
                        req.failBlock!(req)
                    })
                }
            })
            dataTask.resume()
        }
    }
}
